crates/client/eth/src/l1_messaging.rs

```diff
@@ -0,0 +1,497 @@
+        let l1_block_metrics = L1BlockMetrics::register(&prometheus_service.registry()).unwrap();
+
+        // Set up provider
+        let rpc_url: Url = anvil.endpoint().parse().expect("issue while parsing");
+        let provider = ProviderBuilder::new().on_http(rpc_url);
+
+        // Set up dummy contract
+        let contract = DummyContract::deploy(provider.clone()).await.unwrap();
+
+        let core_contract = StarknetCoreContract::new(*contract.address(), provider.clone());
+
+        let eth_client = EthereumClient {
+            provider: Arc::new(provider.clone()),
+            l1_core_contract: core_contract.clone(),
+            l1_block_metrics: l1_block_metrics.clone(),
+        };
+
+        TestRunner { anvil, chain_config, db_service: db, dummy_contract: contract, eth_client }
+    }
+
+    /// Test the basic workflow of l1 -> l2 messaging
+    ///
+    /// This test performs the following steps:
+    /// 1. Sets up test environemment
+    /// 2. Starts worker
+    /// 3. Fires a Message event from the dummy contract
+    /// 4. Waits for event to be processed
+    /// 5. Assert that the worker handle the event with correct data
+    /// 6. Assert that the hash computed by the worker is correct
+    /// 7. TODO : Assert that the tx is succesfully submited to the mempool
+    /// 8. Assert that the event is successfully pushed to the db
+    /// 9. TODO : Assert that the tx was correctly executed
+    #[rstest]
+    #[traced_test]
+    #[tokio::test]
+    async fn e2e_test_basic_workflow(#[future] setup_test_env: TestRunner) {
+        let TestRunner { chain_config, db_service: db, dummy_contract: contract, eth_client, anvil: _anvil } =
+            setup_test_env.await;
+
+        // Start worker
+        let worker_handle = {
+            let db = Arc::clone(&db);
+            tokio::spawn(async move { sync(db.backend(), &eth_client, &chain_config.chain_id).await })
+        };
+
+        let _ = contract.setIsCanceled(false).send().await;
+        // Send a Event and wait for processing, Panic if fail
+        let _ = contract.fireEvent().send().await.expect("Failed to fire event");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+
+        // Assert that event was caught by the worker with correct data
+        // TODO: Maybe add some more assert
+        assert!(logs_contain("fromAddress: 0xae0ee0a63a2ce6baeeffe56e7714fb4efe48d419"));
+
+        // Assert the tx hash computed by the worker is correct
+        assert!(logs_contain(
+            format!("event hash : {:?}", contract.getL1ToL2MsgHash().call().await.expect("failed to get hash")._0)
+                .as_str()
+        ));
+
+        // TODO : Assert that the tx has been included in the mempool
+
+        // Assert that the event is well stored in db
+        let last_block =
+            db.backend().messaging_last_synced_l1_block_with_event().expect("failed to retrieve block").unwrap();
+        assert_ne!(last_block.block_number, 0);
```

Comment:
we should check the event index and nonce as well?

Reply:
done for nonce but event_index is 0 as its the first